### DISCUSSION TITLE
How can I get the brew installed path in macOS's non-interactive ssh `$PATH` variable?

### DISCUSSION THREAD

**Discussion by cdayjr** [Homebrew/discussions#1307] [Everyday usage] [answered]
Trying to use `mosh` to connect to a macOS installation that has `mosh` installed via Homebrew results in the following:

```
$ mosh <host>
zsh:1: command not found: mosh-server
Connection to <host> closed.
/usr/bin/mosh: Did not find mosh server startup message. (Have you installed mosh on your server?)
```

Upon closer inspection I found that `mosh-server` was installed in `/usr/local/bin`:
```
$ which mosh-server
/usr/local/bin/mosh-server
```

Getting the `$PATH` from a bare ssh command returns the following:
```
$ ssh <host> 'echo $PATH'
/usr/bin:/bin:/usr/sbin:/sbin
```

If I ssh into the machine, it runs all of the `.profile` files, etc. and does contain the correct path, but this doesn't happen for `mosh` or the command above which I assume is because it's a non-interactive connection it doesn't run those scripts.

Interestingly, `/etc/paths` does contain the correct path:
```
/usr/local/bin
/usr/bin
/bin
/usr/sbin
/sbin
```

so somehow the `$PATH` there gets generated from something else.

I found a workaround by adding a `.ssh/environment` file that contains the following:
```
PATH=/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin
```

and adding this line to `/etc/ssh/sshd_config` and restarting the `ssh` service:
```
PermitUserEnvironment yes
```

This works, but it doesn't seem ideal; apparently there are security concerns with setting `PermitUserEnvironment`.

Has anyone else ran into this? Is there something that can be done to get that path working correctly, preferably automatically when something like `mosh-server` that needs it to function correctly is installed?

**Comment by the-mgt42** [comment]
Thanks, this worked for me, but I agree I'd like a better solution without using `PermitUserEnvironment`

**Comment by carlocab** [comment]
This might help: https://superuser.com/questions/1484960/set-path-when-ssh-into-macos

Searching for `macos ssh path` might also reveal other solutions.

**Comment by gromgit** [comment, marked as answer]
> I assume is because it's a non-interactive connection it doesn't run those scripts

I don't use `mosh`, but I'm quite sure it creates an interactive connection, else it would be pointless. :)

What `ssh` does when directly invoking a command on the server is to spawn a interactive _non-login_ shell, so `~/.zprofile` will never be sourced. Since you're using `zsh`, the following paragraph from the STARTUP/SHUTDOWN FILES section of the `zsh` man page is relevant:

> Commands are then read from `$ZDOTDIR/.zshenv`.  If the shell is a login shell, commands are read from `/etc/zsh/zprofile` and then `$ZDOTDIR/.zprofile`.  Then, if the shell is interactive, commands are read from `/etc/zsh/zshrc` and then `$ZDOTDIR/.zshrc`.  Finally, if the shell is a login shell, `/etc/zsh/zlogin` and `$ZDOTDIR/.zlogin` are read.

So if you want Homebrew stuff to be available no matter when, I think adding:
```
eval "$(/usr/local/bin/brew shellenv)"
```
to `~/.zshenv` instead of `~/.zprofile` would probably be the best solution.

**Reply by the-mgt42** [reply]
I added
```
export PATH=/usr/local/bin:$PATH
```
to `~/.zshenv` and it works now.

**Reply by cdayjr** [reply]
Thanks! Adding the following to `~/.zshenv` did the trick for me and makes perfect sense.

```zsh
# Homebrew path support
if command -v /usr/local/bin/brew >/dev/null 2>&1; then
  eval "$(/usr/local/bin/brew shellenv)"
fi
```

> I don't use mosh, but I'm quite sure it creates an interactive connection, else it would be pointless. :)

It does eventually over `mosh` itself, but it uses SSH initially in a non-interactive way to spawn a mosh connection after that.

Per https://mosh.org/
> # Same login method.
> Mosh doesn't listen on network ports or authenticate users. The mosh client logs in to the server via SSH, and users present the same credentials (e.g., password, public key) as before. Then Mosh runs the mosh-server remotely and connects to it over UDP.

**Reply by corneliusroemer** [reply]
Note, on ARM macs, the brew prefix has changed from `/usr/local` to `/opt/homebrew`, so you need to add:

```zsh
# Homebrew path support
if command -v /opt/homebrew/bin/brew >/dev/null 2>&1; then
  eval "$(/opt/homebrew/bin/brew shellenv)"
fi
```

to `~/.zshenv`.

**Reply by gromgit** [reply]
A (fairly) portable chunk of shell code, if you have one set of startup scripts across multiple platforms like me:
```bash
homebase=
case "$(uname -sm)" in
  Darwin\ x86_64) homebase=/usr/local;;
  Darwin\ arm64) homebase=/opt/homebrew;;
  Linux\ x86_64) homebase=/home/linuxbrew/.linuxbrew;;
esac
if [[ -z $homebase ]]; then
  echo "NOTE: Homebrew is not supported on this platform" >&2
elif [[ ! -x ${homebase}/bin/brew ]]; then
  echo "NOTE: Homebrew is not installed" >&2
else
  eval "$("${homebase}/bin/brew" shellenv)"
fi
```

**Reply by corneliusroemer** [reply]
@gromgit but you need to put that in different files depending on which shell you use. Maybe you could add another script that automatically adds your chunk of code to a variety of startup scripts: `.zshenv`, `.profile` etc!

**Reply by gromgit** [reply]
That's an exercise for the reader. 😀
